lib/SILOptimizer/Mandatory/DIMemoryUseCollectorOwnership.cpp

```diff
@@ -950,6 +950,31 @@ void ElementUseCollector::collectUses(SILValue Pointer, unsigned BaseEltNo) {
       continue;
     }
 
+    // unchecked_take_enum_data_addr takes the address of the payload of an
+    // optional, which could be used to update the payload. So, visit the
+    // users of this instruction and ensure that there are no overwrites to an
+    // immutable optional. Note that this special handling is for checking
+    // immutability and is not for checking initialization before use.
+    if (auto *enumDataAddr = dyn_cast<UncheckedTakeEnumDataAddrInst>(User)) {
+      // Keep track of the fact that we're inside of an enum. This informs our
+      // recursion that tuple stores should not be treated as a partial
+      // store. This is needed because if the enum has data it would be accessed
+      // through tuple_element_addr instruction. The entire enum is expected
+      // to be initialized before any such access.
+      llvm::SaveAndRestore<bool> X(InEnumSubElement, true);
+      collectUses(enumDataAddr, BaseEltNo);
+      continue;
+    }
```

Comment:
The change compared to the previous fix is that the instruction at line 964 is now added so that accesses to  a tuple element are not considered as a partial store but as a normal store, if the access happens via a pointer to the data of an optional (or enum). This is what happens when tuple elements of an optional tuple are accessed through optional chaining. This logic was borrowed from the `InitEnumDataAddrInst` which I realized already had to handle a similar issue.